mloop/neuralnet.py

```diff
@@ -648,8 +648,8 @@ def fit_neural_net(self, all_params, all_costs):
                 cv_size = int(len(all_params) / 10)
                 train_params = all_params[:-cv_size]
                 train_costs = all_costs[:-cv_size]
-                cv_params = all_params[cv_size:]
-                cv_costs = all_costs[cv_size:]
+                cv_params = all_params[-cv_size:]
```

Comment:
Haha, damn it. Maybe this is why I never had much success with this method!

That said, what do you think about changing to
```
cv_params = all_params[:cv_size]
cv_costs = all_params[:cv_size]
train_params = all_params[cv_size:]
train_costs = all_params[cv_size:]
```
just to save messing around with negative indices?

Reply:
Hmm, that's got me thinking actually. Do you think it would be better to use the most recent parameters in the cross validation set? Generally the later parameters are likely to be closer to the optimal values than the earlier ones. That's the region of parameter space where we'd care most about the effects of overfitting, so I think that's the region of parameter space where the cross validation data points would ideally be.

Put another way, I suspect that if the net is trained mostly on data points near the optimum, then it wouldn't be ideal to assess its "overfitted-ness" at some data points in a potentially far-off region of parameter space.

To avoid negative indices we could still do something like this:
```python
training_fraction = 0.9
split_index = int(training_fraction * len(all_params))
train_params = all_params[:split_index]
train_costs = all_params[:split_index]
cv_params = all_params[split_index:]
cv_costs = all_params[split_index:]
```

I'll implement something like that for now, but I'd be happy to hear your thoughts and wouldn't at all mind changing things.

Reply:
Implemented in 27157d5.

Reply:
I don't have much intuition for this, although I agree that just taking cross validation parameters from the start (like I proposed) probably isn't ideal.

One concern with taking them from the end is that we might go too far the other way, and get a worse fit as a result of not training on any of the parameters closest to the optimum (minimising the cross validation loss will help, but it could be that the minimising fit is still kind of bad).

Still, what you've got definitely seems like a step in the right direction, so I'm happy to stick with that and iterate if necessary once we've seen how it works in action.